src/app/qgsrelationmanagerdialog.cpp

```diff
@@ -15,37 +15,79 @@
 #include "qgsrelationmanagerdialog.h"
 #include "qgsrelationmanager.h"
+#include "qgspolymorphicrelation.h"
 #include "qgsvectorlayer.h"
 
+
+#ifndef SIP_RUN
+class RelationNameEditorDelegate: public QStyledItemDelegate
+{
+  public:
+    RelationNameEditorDelegate( const QList<int> &editableColumns, QObject *parent = nullptr )
+      : QStyledItemDelegate( parent )
+      , mEditableColumns( editableColumns )
+    {}
+
+    virtual QWidget *createEditor( QWidget *parentWidget, const QStyleOptionViewItem &option, const QModelIndex &index ) const
+    {
+      if ( mEditableColumns.contains( index.column() ) )
+        return QStyledItemDelegate::createEditor( parentWidget, option, index );
+
+      return nullptr;
+    }
+  private:
+    QList<int> mEditableColumns;
+};
+#endif
+
 QgsRelationManagerDialog::QgsRelationManagerDialog( QgsRelationManager *relationMgr, QWidget *parent )
   : QWidget( parent )
   , Ui::QgsRelationManagerDialogBase()
   , mRelationManager( relationMgr )
 {
   setupUi( this );
+
+  mRelationsTree->header()->setSectionResizeMode( QHeaderView::ResizeToContents );
+  mRelationsTree->setItemDelegate( new RelationNameEditorDelegate( QList<int>( {0} ), mRelationsTree ) );
+
   connect( mBtnAddRelation, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnAddRelation_clicked );
+  connect( mActionAddPolymorphicRelation, &QAction::triggered, this, &QgsRelationManagerDialog::mActionAddPolymorphicRelation_triggered );
+  connect( mActionEditPolymorphicRelation, &QAction::triggered, this, &QgsRelationManagerDialog::mActionEditPolymorphicRelation_triggered );
   connect( mBtnDiscoverRelations, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnDiscoverRelations_clicked );
   connect( mBtnRemoveRelation, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnRemoveRelation_clicked );
 
   mBtnRemoveRelation->setEnabled( false );
-  connect( mRelationsTable->selectionModel(), &QItemSelectionModel::selectionChanged, this, &QgsRelationManagerDialog::onSelectionChanged );
+  mBtnAddRelation->setPopupMode( QToolButton::MenuButtonPopup );
+  mBtnAddRelation->addAction( mActionAddPolymorphicRelation );
+  mBtnAddRelation->addAction( mActionEditPolymorphicRelation );
+
+  connect( mRelationsTree->selectionModel(), &QItemSelectionModel::selectionChanged, this, &QgsRelationManagerDialog::onSelectionChanged );
 }
 
 void QgsRelationManagerDialog::setLayers( const QList< QgsVectorLayer * > &layers )
 {
   mLayers = layers;
 
   const QList<QgsRelation> &relations = mRelationManager->relations().values();
-
-  const auto constRelations = relations;
-  for ( const QgsRelation &rel : constRelations )
+  for ( const QgsRelation &rel : relations )
   {
+    // the generated relations for polymorphic relations should be ignored,
+    // they are generated when the polymorphic relation is added to the table
+    if ( !rel.polymorphicRelationId().isEmpty() )
+      continue;
+
     addRelation( rel );
```

Comment:
call the private method (see further down)

Reply:
Here it should remain `addRelation`, since we are adding a normal relation.

Reply:
the private method also insert normal relation

Reply:
yes, but using the `addRelationPrivate` is not clear if we "forgot" the parent item, or we wanted to add a normal relation. I read "addRelation" as "addNormalRelation", whether there is a private method is a detail. So for me there should be only two places where `addRelationPrivate` is used, in the `addRelation` and `addPolymorphicRelation` methods. 

If you think it is confusing, I can change that.

Reply:
well, it depends a bit on the comment I left in the public method. If it handles all relation, then we should call the private one here.